Refactor `OllamaService` (timeout fix, model info retrieval, clean code, etc.) + optimized and added test cases including UI

bmartin042503 opened the pull request:
Closes https://github.com/4foureyes/avallamaplanning/issues/78
Closes https://github.com/4foureyes/avallamaplanning/issues/109

### Changes
- Changed `HorizontalAlignment` to left for `SubText` of `ConversationItem` for better readability
- Changed Model Manager button icon
- Refactored and optimized `OllamaService`:
    - Fixed timeout logic by using separate `HttpClient` instances for connection checks and heavier operations (It might potentially solve the following issue: https://github.com/4foureyes/avallamaplanning/issues/111)
    - Refactored downloaded model information retrieval
    - Reorganized all of the code into regions
    - Added detailed documentation
    - etc.
- Removed use of reflection from `OllamaServiceTests.cs` and added a few test cases
- After finishing a download of a model you can see now it's information right away in the Model Manager, and it will cache downloaded models.
- Added UI test cases for `HomeView`
- Optimized test cases and services to use time providers and task delayers (`OllamaService`, `NetworkSpeedCalculator`). So unit tests will run faster and GitHub Actions will pass on those test cases.
- Minor UI and logic changes..

bmartin042503 committed:
refactor and optimize ollamaservice + minor changes

bmartin042503 committed:
timeout fix + minor changes + first ui test

bmartin042503 committed:
test cases + timeprovider for tests + cleaned code

eyeonspringfield commented on avallama.Tests/Mocks/TaskDelayerMock.cs:
Shouldn't this be "Mock" instead of "Fake"?

bmartin042503 replied:
yeah sure, I was uncertain to make a Mock dir and files

eyeonspringfield commented on avallama.Tests/Services/ModelCacheServiceTests.cs:
```diff
@@ -24,11 +24,17 @@ public class ModelCacheServiceTests : IAsyncDisposable
     {
         var tempDbPath = Path.Combine(Path.GetTempPath(), $"avallama_test_{Guid.NewGuid()}.db");
-        _connection = new SqliteConnection($"Data Source={tempDbPath}");
+        var connectionString = $"Data Source={tempDbPath};Pooling=false;Default Timeout=30";
```
Why is `Pooling=false;Default Timeout=30` needed? Also, Default Timeout is default of 30: https://learn.microsoft.com/en-us/dotnet/standard/data/sqlite/connection-strings#default-timeout

bmartin042503 replied:
Oh right, then I'll remove the timeout. I tried to change some settings so it might pass easier on GH actions. Setting `Pooling` to false might help a little, it could be slower but more stable, although I don't have exact evidence of it.

eyeonspringfield commented on avallama.Tests/Services/OllamaServiceTests.cs:
```diff
@@ -56,171 +62,234 @@ public OllamaServiceTests(TestServicesFixture fixture)
 
     [Fact]
-    public async Task CallingOllamaServiceMethod_AfterStart_DoesNotThrow()
+    public async Task StartingOllamaService_WhenMockedRunning_SetsRunning()
```
Nitpick but I don't think the test method name should say "WhenMocked", mocking is just the means to achieving a state where it is running without integrating other components, so it remains a unit test, but testing an explicitly mocked state would mean that the test doesn't test the real functionality.

bmartin042503 replied:
yep I'll change, it was in your commit tho xD https://github.com/4foureyes/avallama/pull/15/changes#diff-c213470a2d709d7446034b263b8042f9031cbd608aa48837d81fca055eb9cfecR116

bmartin042503 replied:
and i didn't notice it either

eyeonspringfield replied:
I was a fool

eyeonspringfield commented on avallama.Tests/Services/OllamaServiceTests.cs:
```diff
@@ -415,4 +498,159 @@ public async Task GenerateMessage_InvalidJson_OnlyShowsValid()
+        {
+            if (s?.Status != null) statusEvents.Add(s.Status);
+        };
+        */
```
Commented out code left in

eyeonspringfield commented on avallama.Tests/Views/HomeViewTests.cs:
```diff
@@ -0,0 +1,337 @@
+    // this is needed so we can override InitializeAsync() to use it for testing
+    // so it won't be necessary to call it again when doing UI testing since OnAttachedToVisualTree will be called
+    // when testing ViewModels calling InitializeAsync() might be necessary still
+    private class TestHomeViewModel(
```
Is there a better way to handle this than to create a separate class for testing? This decreases maintainability (if HomeViewModel changes, this needs to be updated) and introduces code duplication.

bmartin042503 replied:
Yeah probably, I'll need to find a better way to make async initialization then I can change this too. I'll look into it.

eyeonspringfield commented on avallama.Tests/Views/HomeViewTests.cs:
```diff
@@ -0,0 +1,337 @@
+        viewModel.SelectedModelName = "test-model-1:8b";
+
+        // we use Raise and not RaiseAsync cause the delegate in OllamaService returns void not a Task
+        _fixture.OllamaMock.Raise(x => x.ServiceStateChanged += null, new ServiceState(ServiceStatus.Running));
```
This instrumentation logic could be delegated to a helper method

bmartin042503 replied:
yep i'll clean it

eyeonspringfield commented on avallama/Services/OllamaService.cs:
```diff
@@ -11,658 +11,752 @@
+                catch (Exception ex)
+                {
+                    // TODO: proper logging
+                    Console.WriteLine($"Failed to get model info for {downloadedModel.Name}: {ex.Message}");
```
I don't think we should include random Console.WriteLines in production code, even though the proper error handling framework has not yet been set up

bmartin042503 replied:
sure i'll remove console logs, i don't find it fancy either

eyeonspringfield commented on avallama/Services/OllamaService.cs:
```diff
@@ -11,658 +11,752 @@
-    {
-        CurrentServiceStatus = status;
-        CurrentServiceMessage = message;
+        private static void KillProcess(Process? process)
```
I think this could be inlined in `public void Stop()` for simplicity. This seems a little fragmented

bmartin042503 replied:
sure

eyeonspringfield commented on avallama/Utilities/Time/ITaskDelayer.cs:
Could we not leverage the ability to include multiple classes/interfaces in a single file with these time classes? 4 separate files for what is essentially just wrappers seems excessive

bmartin042503 replied:
I was thinking about it too at first, but what should be the filename then? I made 4 files because I think both the interface and the implementation are important in the same way. And if we'd use 2 files only (RealTimeProvider.cs and RealTaskDelayer.cs) they would make interface definitions "hidden". While in other places we can use it more conveniently since, for example IOllamaService and OllamaService is really similar and the interface itself is really small for a single file.

But honestly I don't know either.

eyeonspringfield commented on avallama/Utilities/NetworkSpeedCalculator.cs:
```diff
@@ -3,28 +3,29 @@
+        public NetworkSpeedCalculator(ITimeProvider? timeProvider = null)
         {
-            _stopwatch.Start();
+            // Ha null, akkor az igazi órát használjuk (default behavior)
```
Nitpick, Hungarian comment

bmartin042503 replied:
right, will translate it

eyeonspringfield commented on avallama/ViewModels/HomeViewModel.cs:
```diff
@@ -71,26 +71,22 @@ public ObservableCollection<string> AvailableModels
-    }
+    } = string.Empty;
 
-    // TODO:
-    // Van egy bug amit kicsit nehezen lehet reprodukálni, de épp a 3. üzenetem írtam, generálta volna az új beszélgetés címet az ollama
-    // majd átkattintottam gyors egy másik beszélgetésbe és annak a címébe vitte bele az új címet, hozzáfűzte
-    // és az előző beszélgetés címe eltűnt, nem is tudtam belekattintani, ezt meg lehet csinálni többször, átviszi a generálást máshova
-    // ezt majd vhogy javítani
```
Was this resolved or why was this TODO removed?

bmartin042503 replied:
I removed it because we already made an issue for that, so I thought it might be unnecessary to include in the comments

eyeonspringfield replied:
Oh yeah, that's true

eyeonspringfield commented on avallama/Styles/TemplatedControls/ConversationItem.axaml:
```diff
@@ -76,7 +76,7 @@ Licensed under the MIT License. See LICENSE file for details.
                             Foreground="{DynamicResource OnSurface}"
                             FontSize="12"
-                            HorizontalAlignment="Right"
+                            HorizontalAlignment="Left"
```
I personally don't agree with this alignment change, I think it is more readable if it is on the right. When on the right is is more easily discernible that the text is not part of the title, but additional info.

bmartin042503 replied:
I chose to set it to Left because I thought reading it is more convenient, if the user looks at the title he can see the subtext too, but if its on the right they have to look there. Small thing but yes i can revert it cause I think we both need to agree on UI changes.

eyeonspringfield commented on avallama/ViewModels/ModelManagerViewModel.cs:
```diff
@@ -442,63 +451,21 @@ private void UpdateDownloadSpeedText(object? state)
+            case ServiceStatus.Running:
+                Console.WriteLine("Running..");
+                break;
+            case ServiceStatus.Retrying:
+                Console.WriteLine("Retrying..");
+                break;
+            case ServiceStatus.Stopped or ServiceStatus.Failed:
+                Console.WriteLine("Stopped..");
+                break;
```
Again, I don't think this should be printed to console in production code

bmartin042503 replied:
it was for testing but i forgot to remove

eyeonspringfield reviewed (changes requested):
Tested on Windows, and I have 2 gripes with the current behavior:
1. The startup behavior is still weird. On Windows when starting the app, the dropdown at the top is selectable for 0.5-1 seconds, then it changes to unselectable with the "Loading models" text and flashes a loading text on the screen for a split second, before going back to selectable once models are loaded. I'll send you a video of this in private.

2. There is now a new soft lock situation that can be attained by doing the following:

- Delete `avallama.db`
- Open the app and initialize the model manager library
- Go back to the home view

The app enters a soft locked state where it is "unable to connect" to ollama no matter how many times you retry, and it can't be used until restarted.

This also once happened randomly when I simply navigated to the model manager (when it was already initialized), browsed a few models and navigated back to the home view.

bmartin042503 committed:
refactored homevm + minor changes/fixes

bmartin042503 committed:
soft lock state and retry fix + minor changes

bmartin042503 commented:
### Changes
- Removed console logs and unnecessary comments
- Reverted `HorizontalAlignment` to `Right` for `SubText` within `ConversationItem`
- Refactored and optimized `HomeViewTests`
- Delegated async initialization of ViewModels to AXAML using the new `ViewInitialization` class
- Moved the Ollama server startup `Task` from `HomeViewModel` to `OllamaService`
- Improved Ollama retry logic: added a process check and ensured models are re-initialized upon a `Running` state
- Fixed UI flickering issues and organized `HomeViewModel` code into regions with documentation
- Fixed soft lock state when returning to `HomeView` from `ModelManagerView` and optimized model updates using parallelism to prevent timeouts (socket exhaustion) on Windows

eyeonspringfield reviewed (approved):
Tested on Windows, the brought up issues seemed to have been resolved, everything else works fine.
